### PR TITLE
Service design

### PR DESCRIPTION
**Pull request by katiasuya**

### CODE REVIEW

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

- We're converting audio files not necessarily songs
- We'll use Postgres as a DB, my bad I forgot to tell you
- Let's think about what entity is `converted` really (we can rename endpoints and a table)
- How do I download an original audio file and converted one? Where do I get this info from?
- Statuses

**Comment by katiasuya** [comment]
> * How do I download an original audio file and converted one? Where do I get this info from?

I think we can obtain the place where the user can download his audios from the bucket  with the formula /{User.username}/{Audio.name}.{Audio.format} when needed, so there is no need to save the path in DB, and for me it doesn't make sense either to have the bucket name in the database as it will be always the same for all audios.

**Comment by papahmsolo** [comment]
> /{User.username}/{Audio.name}.{Audio.format}

What happens if I have the same audio file name and type? Also, there's no Object storage and bucket for local development. So we need to know something about how to download the file.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good!

-  Great job grouping endpoints by domain, but `/audios/request` is about a different entity, it's not about audio files, right?
- Let's make `Format` and `Status` enums, we won't need the tables for them then.
- Let's rename `creation` to `created` and also add `updated` for records that can be updated.
- Let's change ids to UUID 
- We still need to have a reference to the image in the storage somewhere in the DB

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

**DB script**
- `updated` is not nullable, it's just the same as `created` upon record creation.
- there could be a problem with time zones, so let's make `timestamp` independent from time zone.
- let's think about what default value we'd like our timestamps to have? 
- is request status nullable?
- probably the most important question: how do I get all my conversion requests? Try to write a query for that. 

**Endpoint**
Given the requirements, we probably won't need as many endpoints, but if you want to implement them I'm okay with this. Just let's think of the endpoints naming, maybe we can come up with more precise names.

Last but not least, don't forget to update the README :)

**Comment by katiasuya** [comment]
> * probably the most important question: how do I get all my conversion requests? Try to write a query for that.

We can do it joining the tables this way: 
request.original_id=audio.id
audio.user_id=user.id

I've already written the queries, they all work fine:)

File: README.md [hunk 1]
```diff
@@ -1 +1,14 @@
...
+
+Audio-converter is a service that exposes a RESTful API to convert WAV to MP3 and vice versa. 
+
+## Architecture Diagram
+
+![diagram](docs/architecture.jpeg)
+
+## DataBase
+
+First, download PostgreSQL server, install it on your system and run it.
```
Comment (papahmsolo) [inline README.md:11]: What version?

File: scripts/schema.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+FOREIGN KEY (user_id) REFERENCES converter."user" (id) ON DELETE CASCADE
+);
+
+CREATE TABLE IF NOT EXISTS converter.request (
+id UUID PRIMARY KEY,
+original_id UUID NOT NULL,
+converted_id UUID,
+created TIMESTAMP WITHOUT TIME ZONE DEFAULT NOW() NOT NULL,
+updated TIMESTAMP WITHOUT TIME ZONE DEFAULT NOW() NOT NULL,
+status status DEFAULT 'queued',
```
Comment (papahmsolo) [inline scripts/schema.sql:48]: Not sure if default works well here, I think we should explicitly set the status. but the status cannot be NULL

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool!

- [x] `POST /requests/{id}` do we know the ID before we create the request ?:)
- [x] It's okay for now but later on, we'll need a way to authenticate a user, so we'll have to add it.
- [ ] Which endpoint is responsible for the conversion?

File: scripts/schema.sql (hunk 2 above)
Comment (katiasuya) [reply scripts/schema.sql:48]: I put not null in the architecture diagram but not in the code, my bad :(

**Comment by katiasuya** [comment]
> * [ ]  Which endpoint is responsible for the conversion?
I think the conversion will be triggered in POST method of /requests endpoint.

File: scripts/queries.sql [hunk 3]
```diff
@@ -0,0 +1,93 @@
+   -- to create a user
+   
+    INSERT INTO converter."user" (username, password) VALUES
+    ('username', 'password');
+
+   -- to upload an audio
+   
+    INSERT INTO converter.audio (user_id, name, format, location) VALUES
+    ('61e72557-e5af-4bc2-b19e-b1e4c7820d14','song name', 'MP3', 'location');
```
Comment (papahmsolo) [inline scripts/queries.sql:9]: You may want to obtain the id of the newly created record, you'll need this in the request record

File: scripts/queries.sql [hunk 4]
```diff
@@ -0,0 +1,93 @@
+   -- to create a user
```
Comment (papahmsolo) [inline scripts/queries.sql:1]: You'll also need to select the user for login

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

- [x] We can discuss the endpoint offline
- [x] We need to create a DB user in Postgres for the services, as we cannot use admin credentials for that, and it's better be in the script as well.

File: scripts/queries.sql (hunk 4 above)
Comment (katiasuya) [reply scripts/queries.sql:1]: Do you mean comparing passwords for a particular username?

File: scripts/schema.sql [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+WHERE NOT EXISTS (SELECT FROM pg_database WHERE datname = 'audioconverter');
+\gexec
+
+DO $$
+BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname='acuser') THEN
+        CREATE USER acuser WITH ENCRYPTED PASSWORD 'AcUser!';
+        GRANT ALL PRIVILEGES ON DATABASE audioconverter TO acuser;
+    END IF;
+END$$;
```
Comment (papahmsolo) [inline scripts/schema.sql:11]: Storing credentials in the repo is a huge security breach, you should rather prompt to enter the user/password after you run the script.

File: scripts/queries.sql [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+   
+    INSERT INTO converter."user" (username, password) VALUES
+    ('username', 'password');
+
+   -- to log in the user getting the password
+
+    SELECT u.password
+    FROM converter."user" u
+    WHERE username = 'aaa';
```
Comment (papahmsolo) [inline scripts/queries.sql:10]: Not really an issue, but it does not make sense to use the alias in such short queries.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

File: README.md [hunk 7]
```diff
@@ -1 +1,15 @@
...
+## Architecture Diagram
+
+![diagram](docs/architecture.jpeg)
+
+## DataBase
+
+First, download PostgreSQL server of the version 13.x, install it on your system and run it.
+
+To create a PostgreSQL user, database, schema and tables needed for the service, run 
+`scripts/create_db.sh`.
```
Comment (papahmsolo) [inline README.md:14]: You'll probably get an error if you run the script from the root folder, you need to `cd scripts` first

File: scripts/create_db.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+echo 'Please, enter username:'
+read  DB_USERNAME
+echo 'Please, enter password:'
+read  DB_PASSWORD
+
+psql \
+	--variable=user_var=$DB_USERNAME \
+	--variable=password_var=$DB_PASSWORD \
+	-f schema.sql
```
Comment (papahmsolo) [inline scripts/create_db.sh:11]: - I think you need to run all the command as the DB admin, by default it asked me for a domain user password but should ask for `postgres` user password
- new line at the end of the file :)

File: scripts/create_db.sh [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+echo 'Please, enter username:'
+read  DB_USERNAME
+echo 'Please, enter password:'
+read  DB_PASSWORD
```
Comment (papahmsolo) [inline scripts/create_db.sh:6]: `-s` - do not echo password coming from a terminal

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool!

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**